include/ofi_hmem.h

```diff
@@ -120,6 +120,7 @@ struct ofi_hmem_ops {
 			     size_t *base_length);
 	bool (*is_ipc_enabled)(void);
 	int (*get_ipc_handle_size)(size_t *size);
+	bool (*is_dev_reg_enabled)(void);
```

Comment:
Hi Thomas,

Could you explain the use-case for this and why the current dev_reg/unreg APIs will not work?

The use-case that dev_reg/unreg enables is for providers to cache a dev_reg handle with their memory registration. During memory registration, if dev_reg fails, the expectation is the provider implements a fallback. I would expect the fallback to be using the non-dev_reg memcpy APIs which could hurt performance.

Reply:
sure, so all this came from the need of adding `cuda gdrcopy` to `rxm` (as done already in `efa`) when doing injection/inlining for send/recv. With cuda, `gdrcopy` is used through the `dev_reg` features and is conditionally registered at mr time (with cuda_is_gdrcopy_enabled):

https://github.com/ofiwg/libfabric/blob/c37c44e5b485b5185de96c53453a7249ca3fd672/prov/rxm/src/rxm_domain.c#L602-L613

https://github.com/ofiwg/libfabric/blob/9f604d658277c3a4851506327cb7a75572622767/prov/efa/src/efa_mr.c#L253-L256

one of the comment in [this PR](https://github.com/ofiwg/libfabric/pull/9454#discussion_r1369355045) highlighted that it doesn't make sense to keep the conditional in cuda as the `dev_reg` feature should be used by all the iface (rocr, cuda, ze, etc), hence the present draft PR.
It is still unclear whether or not it should be enabled by default for rocr iface. From your answer I extrapolate that it should be done at every `mr`?

Reply:
I guess I question if this is_dev_reg_enabled needed. I think the code in https://github.com/ofiwg/libfabric/blob/c37c44e5b485b5185de96c53453a7249ca3fd672/prov/rxm/src/rxm_domain.c#L602-L613 could be implemented as follows.

```
 	int ret = ofi_hmem_dev_register(rxm_mr->iface, attr->mr_iov->iov_base, attr->mr_iov->iov_len, (uint64_t *) &rxm_mr->hmem_handle);
	if (ret) {
		rxm_mr->hmem_flags = 0x0; 
 		rxm_mr->hmem_handle = NULL; 
	} else if (rxm_mr->iface == FI_HMEM_CUDA) {
		rxm_mr->hmem_flags = OFI_HMEM_DATA_GDRCOPY_HANDLE;
	}
```

Reply:
I don't have the history, but from the code one of the reason is that`ofi_hmem_dev_register` does not return an error if `FI_HMEM_CUDA_USE_GDRCOPY=0`.
@wenduwan @shijin-aws do you have more input?

Also, why only setting `OFI_HMEM_DATA_GDRCOPY_HANDLE` for cuda?
From Edgar's answer it would be useful for AMD as well.

Reply:
Making OFI_HMEM_DATA_GDRCOPY_HANDLE more generic for all HMEM seems to be the correct change.

Reply:
@thomasgillis  Thanks for the heads up. I agree with @iziemba wrt if `ofi_hmem_dev_register` fails then the caller should have a fallback.

EFA only attempts `ofi_hmem_dev_register` if gdrcopy is enabled and available https://github.com/ofiwg/libfabric/blob/9f604d658277c3a4851506327cb7a75572622767/prov/efa/src/efa_mr.c#L253

Are you suggesting that we change `ofi_hmem_dev_register` so that it returns error if gdrcopy is disabled? I think that is also fine and backward compatible for EFA.

Reply:
I think making it fail would be more iface agnostic and help to comply to the if(error) don't do it strategy. I will try to make the changes asap then

Reply:
@wenduwan I have done the changes for the common part but only replaced the naming in EFA.
Is this the right way to do things to not break anything in EFA?

Reply:
Please make the efa change in the same commit where you renamed the macro. We shouldn't have a commit that cannot be compiled